cmd/controller/app/controller.go

```diff
@@ -63,89 +66,145 @@ const controllerAgentName = "cert-manager"
 //and following discussion: https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408500629
 const resyncPeriod = 10 * time.Hour
 
-func Run(opts *options.ControllerOptions, stopCh <-chan struct{}) {
+func Run(opts *options.ControllerOptions, stopCh <-chan struct{}) error {
 	rootCtx := cmdutil.ContextWithStopCh(context.Background(), stopCh)
+	rootCtx, cancelContext := context.WithCancel(rootCtx)
+	defer cancelContext()
+	g, rootCtx := errgroup.WithContext(rootCtx)
 	rootCtx = logf.NewContext(rootCtx, nil, "controller")
 	log := logf.FromContext(rootCtx)
 
-	ctx, kubeCfg, err := buildControllerContext(rootCtx, stopCh, opts)
+	ctx, kubeCfg, err := buildControllerContext(rootCtx, opts)
 	if err != nil {
 		log.Error(err, "error building controller context", "options", opts)
-		os.Exit(1)
+		return err
 	}
 
 	enabledControllers := opts.EnabledControllers()
 	log.Info(fmt.Sprintf("enabled controllers: %s", enabledControllers.List()))
 
-	metricsServer, err := ctx.Metrics.Start(opts.MetricsListenAddress, opts.EnablePprof)
+	ln, err := net.Listen("tcp", opts.MetricsListenAddress)
 	if err != nil {
 		log.Error(err, "failed to listen on prometheus address", "address", opts.MetricsListenAddress)
-		os.Exit(1)
+		return err
 	}
+	server := ctx.Metrics.NewServer(ln, opts.EnablePprof)
 
-	var wg sync.WaitGroup
-	run := func(_ context.Context) {
-		for n, fn := range controller.Known() {
-			log := log.WithValues("controller", n)
+	g.Go(func() error {
+		<-rootCtx.Done()
+		// allow a timeout for graceful shutdown
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
 
-			// only run a controller if it's been enabled
-			if !enabledControllers.Has(n) {
-				log.V(logf.InfoLevel).Info("not starting controller as it's disabled")
-				continue
+		if err := server.Shutdown(ctx); err != nil {
+			return err
+		}
+		return nil
+	})
+	g.Go(func() error {
+		log.WithValues("address", ln.Addr()).V(logf.InfoLevel).Info("listening for connections on")
+		if err := server.Serve(ln); err != http.ErrServerClosed {
+			return err
+		}
+		return nil
+	})
+
+	elected := make(chan struct{})
+	if opts.LeaderElect {
+		g.Go(func() error {
+			log.V(logf.InfoLevel).Info("starting leader election")
+			leaderElectionClient, err := kubernetes.NewForConfig(rest.AddUserAgent(kubeCfg, "leader-election"))
+			if err != nil {
+				log.Error(err, "error creating leader election client")
+				return err
 			}
 
-			// don't run clusterissuers controller if scoped to a single namespace
-			if ctx.Namespace != "" && n == clusterissuers.ControllerName {
-				log.V(logf.InfoLevel).Info("not starting controller as cert-manager has been scoped to a single namespace")
-				continue
+			errorCh := make(chan error, 1)
```

Comment:
Does this channel need to be cleaned up on exit?
```suggestion
			errorCh := make(chan error, 1)
			defer close(errorCh)
```